### PR TITLE
fix the matcher for graph with unused placeholder

### PR DESCRIPTION
**Pull request by cccclai**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90792

it's possible that some placeholders are not used but they won't be eliminated. Fix the logic when we do graph matching for graph with unused input

Differential Revision: [D42011594](https://our.internmc.facebook.com/intern/diff/D42011594/)

Differential Revision: [D42011594](https://our.internmc.facebook.com/intern/diff/D42011594)

### CODE REVIEW

File: test/fx/test_pass_infra.py [hunk 1]
```diff
@@ -232,3 +234,18 @@ def pass_fail(graph_module):
...
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x, y):
+                return x
+
+        m = MyModule()
+        graph_module: torch.fx.GraphModule = symbolic_trace(m)
+        is_matched = compare_graphs(
+            graph_module.graph, graph_module.graph
```
Comment (mergennachin) [inline test/fx/test_pass_infra.py:249]: Can you add more test cases? This one is only comparing with itself.

**Review by mergennachin — COMMENTED** [review]
Also please fix "line too long" lint errors

File: torch/fx/passes/utils/matcher_utils.py [hunk 2]
```diff
@@ -304,6 +306,19 @@ def backtracking(anchor_index, match):
                 match = copy.copy(saved_match)
 
         match = InternalMatch(anchors=self.pattern_anchors)
+
+        # For the unused place holders, if the number of them are the same between pattern and graph,
+        # it's considered as a match and we will add them to the match.nodes_map.
+        graph_placeholder_nodes = [n for n in graph.nodes if n.op == "placeholder"]
+        unused_graph_placeholder_nodes = [placeholder_node for placeholder_node in graph_placeholder_nodes if len(placeholder_node.users) == 0]
+
+        if len(self.unused_pattern_placeholder_nodes) == len(unused_graph_placeholder_nodes):
```
Comment (SherlockNoMad) [inline torch/fx/passes/utils/matcher_utils.py:315]: I am afraid this is not a general rule that we can accept. 
A pattern graph with unused placeholder is not a well-defined pattern. How should the unused placeholder be interpreted in this case? As it has no consumer, it's effectively a dangling argument can be match with any tensor in the target graph. 
For this very reason, we disallow pattern graph with unused placeholder. If we are missing an assert statement to check this, I should add it.

**Review by SherlockNoMad — CHANGES_REQUESTED** [review]
Sorry, pattern graph with unused placeholder is not a well defined pattern.

**Comment by cccclai** [comment]
abandon as we don't need it now.

**Pull request closed without merging**